datalog/solver.go

```diff
@@ -0,0 +1,136 @@
+package datalog
+
+import (
+	"fmt"
+)
+
+// SolvSafeForWorldError allows to make an error that will not cause
+// failed state while Authoriser reconciles the facts using rules
+type SolvSafeForWorldError struct {
+	errMsg string
+}
+
+func (e SolvSafeForWorldError) Error() string {
+	return e.errMsg
+}
+
+func (e SolvSafeForWorldError) Is(target error) bool {
+	if _, ok := target.(*SolvSafeForWorldError); ok {
+		return true
+	}
+	return false
+}
+
+type Solver struct {
+	predicates  []Predicate
+	expressions []Expression
+	allFacts    *FactSet
+	predToFacts [][]*Fact
+	solutions   []MatchedVariables
+	symDeb      SymbolDebugger
+	syms        *SymbolTable
+}
+
+// NewSolver prepeares a Solver instance, will return an error if there are no matching Facts for a Predicate
+func NewSolver(predicates []Predicate, expressions []Expression, allFacts *FactSet, syms *SymbolTable) (*Solver, error) {
+	c := &Solver{
+		predicates:  predicates,
+		expressions: expressions,
+		allFacts:    allFacts,
+		symDeb:      SymbolDebugger{syms},
+		syms:        syms,
+	}
+
+	c.predToFacts = make([][]*Fact, len(predicates))
```

Comment:
why is this created here instead of doing it in `Solve`, where the elements could be consumed as they are generated?

Reply:
ah, disregard, it's because solve is the recursive part

Reply:
so here it creates the list of matching facts, but only depending on static elements, and matches variables in `Solve`, right?